### DISCUSSION TITLE
[feature request] option to use controller tool and base data instead of hardcoding

### DISCUSSION THREAD

**Discussion by lin-ycv** [visose/Robots#154] [Q&A] [answered]
Currently robots hard codes tool and base data as frames then defines the tool and base as those frame
```
$TOOL = DefaultGHTool
$BASE = DefaultFrame
```

But this will cause the controller to do load calculations, and will throw errors and warning for overload and underload when the weight and centre of mass is not configured correctly

if we use the data from the controller, we can disable load verifications in the tool calibration menu
generally this would be a bad idea, but sometimes the weight of the tool changes because the tool weight includes material for additive manufacturing, as the program proceeds, weight becomes lighter.

I've tried to use
```
$TOOL = TOOL_DATA[16]
$BASE = BASE_DATA[32]
```
but it doesn't seems to work properly, it's just importing the values, but not the ignore load setting

**Comment by lin-ycv** [comment]
similar request to https://github.com/visose/Robots/discussions/147
but with different justifications

**Comment by visose** [comment]
To take into account decrease in weight, you can create multiple tools with same TCP but different weight, then assign them along the toolpath to match the weight it's holding at that point.

Better yet is to avoid having the tool hold the bulk of the weight of the material. If not in a fixed position, at least attach it to one of the larger joints.

**Comment by visose** [comment]
> I've tried to use
> 
> ```
> $TOOL = TOOL_DATA[16]
> $BASE = BASE_DATA[32]
> ```
> 
> but it doesn't seems to work properly, it's just importing the values, but not the ignore load setting

You also need to remove or replace any lines starting with $LOAD

**Comment by visose** [comment, marked as answer]
This feature has been added in version 1.4, please test to see if there are any issues. 

Now you can add a `number` attribute to a `Tool` element to load a tool from the controller's TOOL_DATA array. This is only supported in KUKA.
```	
<Tool name="ToolName" number="1">
    <Tcp x="0" y="0" z="0" q1="1" q2="0" q3="0" q4="0" />
    <Mass weight="20" x="5" y="-100" z="100"/>
</Tool>
```
If the tool is predefined as a variable in the controller, you can use `useController` attribute. This is supported on all robot manufacturers, including KUKA. The tool name should match the variable name in the controller.
```
<Tool name="ToolName" useController="true">
    <Tcp x="0" y="0" z="0" q1="1" q2="0" q3="0" q4="0" />
    <Mass weight="20" x="5" y="-100" z="100"/>
</Tool>
```

Now you can also define frames in the robot library and it use the `number` and `useController` attributes mentioned above. 

```
<Frame name="FrameName" number="1">
    <Base x="0" y="0" z="0" q1="1" q2="0" q3="0" q4="0" />
    <Coupling group="0" mechanism="0" />
</Frame>
```
The `Coupling` element is optional, and can be removed if the frame is static.

> `useController` does not work with Frame elements in UR.
> The UR generated script doesn't include frame variables because UR doesn't have a frame equivalent. The transformation is precalculated, and the result baked into the generated script. You can, though, generate UR code that does this transformation math in the controller, so I'm thinking of defining the frame as a pose variable and generate code that calculates this transformation in the controller on every target.
